agents/cluster_methods_mixin.py

```
@@ -101,6 +126,47 @@ def _build_cluster_string(
+        prompt_overhead_tokens = prompt_overhead_chars / ModelCapabilities.CHARS_PER_TOKEN
+        available_tokens = (ctx.input_tokens - OUTPUT_HEADROOM_TOKENS - prompt_overhead_tokens) * CONTEXT_MARGIN
+        char_budget = int(available_tokens * ModelCapabilities.CHARS_PER_TOKEN)
+        if char_budget <= 0:
+            msg = (
+                f"Prompt overhead ({prompt_overhead_chars} chars) consumes the entire agent input "
+                f"window ({ctx.input_tokens} tokens); no room for cluster renderings."
+            )
+            logger.error("[CFG skip planner] %s", msg)
+            raise ContextBudgetExceededError(msg)
+
+        langs_with_clusters = [l for l in programming_langs if cluster_results.get(l)]
```

Ruff (0.15.10)

[error] Ambiguous variable name: `l`

(E741)